### DISCUSSION TITLE
Spec discussion: Give users better control of catalog caching for taps

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#6757] [Spec Discussions]
As raised here:

- https://github.com/meltano/meltano/issues/2848
- And [in slack](https://meltano.slack.com/archives/CKHP6G5V4/p1663216549001279?thread_ts=1663205560.565299&cid=CKHP6G5V4)

## Requirement overview

Some taps generate volatile schema, where the upstream source cannot be considered a closed system or a governed/stable resource. For instance, as a non-admin user on a SQL Server database, I might reasonably assume that my catalog will not change from one minute to the next, but on any given day new tables could be added and existing tables could have their schemas modified - all without notification.

## Options

Users need a way to do one or more of the following:

1. explicitly force a refresh of a catalog, 
3. explicitly invalidate the catalog cache,
2. define an expiration period for a tap's catalog (for instance 12 or 24 hours), after which time the catalog will be considered stale, and/or 
4. specify that a tap's catalog should simply not be cached.

These are all subtly different takes on the same core problem.

## Best workaround as of today

As of today, the only real option is for the user to manually delete the contents of the catalog, for instance by running something similar to: `rm .meltano/run/tap-name/*`.

This is problematic because it is cumbersome but also because we don't want users to have to understand or navigate the internal `.meltano` directory by hand.

**Comment by aaronsteers** [comment]
## Option 1: Explicitly forcing the refresh of a catalog

A new `catalog` command with the subcommand `update` (or `refresh`, or similar) could explicitly rerun discovery.

```
# refresh the specific extractor's catalog before running EL
meltano catalog update tap-name
meltano run tap-name target-name
```

Or:

```
# refresh all catalogs
meltano catalog update --all
meltano run tap-name target-name
```

**Reply by edgarrmondragon** [reply]
The `select` command already works as an interface to the catalog, so perhaps the new sub-command can be added there? My concern is that catalog functionality sprawls across multiple methods.

**Reply by aaronsteers** [reply]
This is a good point and I want to give it some more thought... I wonder if `select` would make sense refactored as a subcommand of `meltano catalog`. 

Adapting slightly from https://docs.meltano.com/reference/command-line-interface#select:

```
meltano catalog select [--list] [--all] <tap_name> [ENTITIES_PATTERN] [ATTRIBUTE_PATTERN]
```

**Reply by aaronsteers** [reply]
@edgarrmondragon - While in the long run, I like the idea of everything catalog related living under a top-level `meltano catalog` command, this is not going to happen fast enough.

I logged an issue following on your proposal here:

- #6919

**Comment by aaronsteers** [comment]
## Option 2: Explicitly forcing the invalidation of a catalog

A new `catalog` command with the subcommand `delete` (or `clear`, `remove`, `rm`, or similar) could explicitly delete the cached catalog.

```bash
# Delete the specific extractor's catalog before running EL
meltano catalog clear tap-name
# Invoking the tap will force a catalog refresh
meltano run tap-name target-name
```

Or:

```bash
# Delete all catalogs
meltano catalog update --all
# Invoking the tap will force a catalog refresh
meltano run tap-name target-name
```

**Comment by aaronsteers** [comment]
## Option 3: Specify max time-to-live for a catalog cache

```console
meltano config tap-name set catalog_max_age '2 hr'
```

```yml
plugins:
  extractors:
  - name: tap-name
    catalog_max_age: 2 hr
    config:
      // ...
```

**Reply by visch** [reply]
Option 3 and Option 3++ are my favorites :)

**Comment by aaronsteers** [comment]
## Option 4: Specify a tap's catalog should not be cached

```console
meltano config tap-name set _catalog_caching disabled
```

```yml
plugins:
  extractors:
  - name: tap-name
    catalog_caching: disabled
    config:
      // ...
```

**Reply by visch** [reply]
Option 3 and Option 3++ are my favorites :)

**Reply by aaronsteers** [reply]
Oops - I had two "option 3s" 😆 . Now this one is correctly numbered as "option 4". 👍

**Reply by edgarrmondragon** [reply]
I like this one too. It probably has to be created as an _extra_ under the hood, so the variable would have to prefixed with `_` in the command:

```
meltano config tap-name set _catalog_caching disabled
```

**Reply by aaronsteers** [reply]
Thanks, @edgarrmondragon. I've updated the above to reflect proper naming convention as an `extra` with preceeding underscore: `_catalog_caching` instead of `catalog_caching`.

**Reply by aaronsteers** [reply]
On thinking about this more, we technically _do_ have to cache the file - but presumably we'd delete it / clean it up before and after each invocation? UPDATE: Or store it at a path specific to the particular execution, not global to the tap.

```consoole
meltano config tap-name set _catalog_caching disabled

# This invokes `--discover` and cleans up before and after execution
meltano select tap-name --list

# This invokes `--discover` again, cleaning up the catalog before and after execution
meltano run tap-name target-name

# This invokes `--discover` again as well
meltano invoke tap-name --dump=catalog
```

~~We could run into a little bit of trouble if the tap is invoked twice at the same time, but this is already something we check for - at least for test runs.~~

Probably a better approach than deleting the file afterward is (when caching is disabled) to store the file with a subfolder path or file name specific to that execution. This approach is safer than simply deleting afterwards, since in theory multiple simultaneous invocations could each have their own run-results subfolder and they would not impact each other.

**Reply by visch** [reply]
@aaronsteers  could copy what happens right now with `config` files where we add a UUID so we don't have an issue with parallelization. So it'd be something like `properties-{{uuid}}.json`

**Reply by aaronsteers** [reply]
@visch - Definitely! Yeah, I think that would work!

**Reply by aaronsteers** [reply]
Related issue came up today:

Having some means of disabling caching also seems like a necessary safety measure to unblock :

- https://github.com/meltano/meltano/issues/6770

**Reply by pnadolny13** [reply]
This came up in slack again today https://meltano.slack.com/archives/C01TCRBBJD7/p1680249669517619
